### DISCUSSION TITLE
Modeling "White Noise" in PyCaret

### DISCUSSION THREAD

**Discussion by ngupta23** [pycaret/pycaret#1837] [FAQs]
# Background

In a **[previous FAQ](https://github.com/pycaret/pycaret/discussions/1800)**, we discussed what "White Noise" is, why it is important to diagnose it, and how you can do that with the `pycaret`. If you have not read it yet, we would recommend doing so now since this FAQ builds on top of the previous one. 

We also mentioned that the best way to model this is using a `Grand Means` model where the prediction is simply the mean value of the data used to train the model. This is the "best" estimate of any future prediction since the data does not contain any signal. Now let's see how we can model white noise with `pycaret`.

Note that this FAQ is also available as a **[notebook here](https://gist.github.com/ngupta23/b63c7480e3eb63dc30ebac483177f4d1)** so you can run the code yourself.

# Modeling White Noise with `pycaret`

Luckily `pycaret` provides a out of the box `Grand Means` model which is enabled when seasonal period is set to 1. 

## Modeling

```python
#### Install library (one time) ----
!pip install pycaret-ts-alpha -U

#### Import libraries ----
from pycaret.datasets import get_data
from pycaret.time_series import TSForecastingExperiment

#### Get the data ---
y = get_data("1", folder="time_series/white_noise")

#### Setup the experiment ----
exp = TSForecastingExperiment()
exp.setup(data=y, fh=30, seasonal_period=1, session_id=42)

#### Model as a `Grand Means` model ----
model = exp.create_model("grand_means")
```

## Model Fit Assessment
```python
#### Passing an estimator to `check_stats` applies the test to the model residuals
exp.check_stats(model, "white_noise")
```
![image](https://user-images.githubusercontent.com/33585645/141292809-9efc4ef3-e596-4aa7-a861-c2c7c656841e.png)

The residuals from the model fit are consistent with white noise as well. Hence the fit seems to be good!

```python
#### Passing an estimator to `plot_model` applies the plot to the model residuals
exp.plot_model(model, "diagnostics")
```
![image](https://user-images.githubusercontent.com/33585645/141292877-763876ba-3cb5-4916-83b8-85a542ed4d2b.png)

The Grand Means Forecaster Residuals show no significant lags and the residuals are normally distributed with a mean of ~ 0. All this is indicative of a good fit.

## Forecast

```python
#### Check the forecasts ----
exp.plot_model(model)
```

![image](https://user-images.githubusercontent.com/33585645/141292979-9312c7be-71dd-48e6-9909-a40df475338c.png)

The forecast is a mean value of the dataset as expected.

But this is just one dataset. How do we know that the Grand Means model is truly the best model for white noise? Once way to get more confidence is to benchmark the Grand Means model against other models using several white noise datasets. Let's see how we can do this using `pycaret`.

# Benchmark | Is Grand Means the best model for white noise? 

## Step 1: Get the datasets for benchmarking 
For benchmarking the results, we will use the datasets that come with pycaret. There are about ~500 white noise datasets available. In order to speed up the process, we will limit it to ~ 100 (94 to be precise) datasets. This gives us a good sample size to make some conclusions.

NOTE: The code for benchmarking is in the notebook linked above. For brevity, I will not repeat it completely here.

## Step 2: Build models for all datasets 
In order to benchmark, we will build all the available models on all the datasets and then analyze the results. Again, in order to speed up the process, I removed some of the slower models, but we are still benchmarking over 20 models per dataset.

## Step 3: Develop Metric

In order to compare the results, we add a `relative_metric` value. For our metric of choice (e.g. MAE), this indicates how much % above or below all the models are compared to the best model for a particular dataset. For example, if the best model for a particular dataset has a MAE of 10 and another model has an MAE of 11, then the relative metric for the second model will be 11/10 or 1.1. The relative metric for the best model will be 1.

For metrics where a lower value is better (such as MAE), lower value of `relative_metric` is better and vice_versa. The best model always has a `relative_metric` of 1 irrespective of the directionality of the metric of choice.

## Step 4: Analyze Results

```python
#### Which one is the most accurate for type of dataset (white noise) ----
groupby_cols = "model"
agg_dict = {'relative_metric': ['mean'], 'TT (Sec)': ['mean']}
sort_cols = ("relative_metric", "mean")

most_accurate = summarize(all_results, groupby_cols=groupby_cols, agg_dict=agg_dict, sort_cols=sort_cols)
most_accurate
```

![image](https://user-images.githubusercontent.com/33585645/141292403-40fddab7-ffee-4ada-8265-0378e1108ec7.png)

```python
#### Which one is the fastest for type of dataset (white noise) ----
groupby_cols = "model"
agg_dict = {'relative_metric': ['mean'], 'TT (Sec)': ['mean']}
sort_cols = ("TT (Sec)", "mean")

fastest = summarize(all_results, groupby_cols=groupby_cols, agg_dict=agg_dict, sort_cols=sort_cols)
fastest
```

![image](https://user-images.githubusercontent.com/33585645/141292458-a6231bf4-bf7e-4a38-b469-1c262f91d90c.png)

## Summary 
So the Grand Means model has the best metric (within 0.7% of the best model on average) on all these datasets. Plus is one of the fastest to model which could matter when scaling the modeling process to several thousands of time series! So next time, you diagnose your dataset as white noise, save yourself some time and go with a "Grand Means" model!

That's it for this FAQ. Happy forecasting!

**Comment by moezali1** [comment]
PyCaret 4.0 sunset the legacy time-series module (sktime/prophet integration), so this FAQ no longer maps to a supported code path.

If you still need time-series forecasting, use sktime / prophet / statsforecast directly — those projects have surpassed what the PyCaret wrapper offered.

For 4.0 docs, see https://pycaret.org/docs and the cleanup notice at https://github.com/pycaret/pycaret/issues/4177.

Closing as outdated.
